diff --git a/hnswlib/hnswalg.h b/hnswlib/hnswalg.h
index e269ae6..370ead4 100644
--- a/hnswlib/hnswalg.h
+++ b/hnswlib/hnswalg.h
@@ -1,5 +1,5 @@
 #pragma once
-
+#include <set>
 #include "visited_list_pool.h"
 #include "hnswlib.h"
 #include <atomic>
@@ -9,6 +9,7 @@
 #include <unordered_set>
 #include <list>
 #include <memory>
+#include <tuple>
 
 namespace hnswlib {
 typedef unsigned int tableint;
@@ -168,6 +169,12 @@ class HierarchicalNSW : public AlgorithmInterface<dist_t> {
             return a.first < b.first;
         }
     };
+    struct CompareTupleByFirst {
+        constexpr bool operator()(std::tuple<dist_t, tableint,int> const& a,
+            std::tuple<dist_t, tableint,int> const& b) const noexcept {
+            return std::get<0>(a)<std::get<0>(b);
+        }
+    };
 
 
     void setEf(size_t ef) {
@@ -439,6 +446,103 @@ class HierarchicalNSW : public AlgorithmInterface<dist_t> {
         return top_candidates;
     }
 
+    template <bool bare_bone_search = true, bool collect_metrics = false>
+    std::priority_queue<std::pair<dist_t, tableint>, std::vector<std::pair<dist_t, tableint>>, CompareByFirst>
+    searchBaseLayerSTUntilAllGroundTruthFound(
+        tableint ep_id,
+        const void *data_point,
+        size_t k,
+        std::set<labeltype>& ground_truth_labels,
+        std::vector<int>& hop_id_list,
+        int& iter_count,
+        int& dist_count,
+        BaseFilterFunctor* isIdAllowed = nullptr,
+        BaseSearchStopCondition<dist_t>* stop_condition = nullptr) const {
+        VisitedList *vl = visited_list_pool_->getFreeVisitedList();
+        vl_type *visited_array = vl->mass;
+        vl_type visited_array_tag = vl->curV;
+        bool all_gt_found = false;
+        int gt_found_count = 0;
+        std::priority_queue<std::pair<dist_t, tableint>, std::vector<std::pair<dist_t, tableint>>, CompareByFirst> top_candidates;
+        // std::priority_queue<std::pair<dist_t, tableint>, std::vector<std::pair<dist_t, tableint>>, CompareByFirst> candidate_set;
+        std::priority_queue<std::tuple<dist_t, tableint, int>, std::vector<std::tuple<dist_t, tableint, int>>, CompareTupleByFirst> candidate_set;
+        
+        if (bare_bone_search || 
+            (!isMarkedDeleted(ep_id) && ((!isIdAllowed) || (*isIdAllowed)(getExternalLabel(ep_id))))) {
+            char* ep_data = getDataByInternalId(ep_id);
+            dist_t dist = fstdistfunc_(data_point, ep_data, dist_func_param_);
+            top_candidates.emplace(dist, ep_id);
+            if (ground_truth_labels.count(getExternalLabel(ep_id))) {
+                gt_found_count++;
+                all_gt_found = (gt_found_count == k);
+            }
+            if (!bare_bone_search && stop_condition) {
+                stop_condition->add_point_to_result(getExternalLabel(ep_id), ep_data, dist);
+            }
+            candidate_set.emplace(-dist, ep_id,0);
+        } else {
+            candidate_set.emplace(-std::numeric_limits<dist_t>::max(), ep_id,0);
+        }
+
+        visited_array[ep_id] = visited_array_tag;
+        all_gt_found = (gt_found_count == k);
+
+        while (!candidate_set.empty()&& !all_gt_found) {
+            std::tuple<dist_t, tableint, int> current_node_pair = candidate_set.top();
+            dist_t candidate_dist = -std::get<0>(current_node_pair);
+            // dist_t candidate_dist = -current_node_pair.first;
+            candidate_set.pop();
+            iter_count++;
+            // tableint current_node_id = current_node_pair.second;
+            tableint current_node_id = std::get<1>(current_node_pair);
+            int cur_hop_id=std::get<2>(current_node_pair);
+            hop_id_list.emplace_back(cur_hop_id);
+            int *data = (int *) get_linklist0(current_node_id);
+            size_t size = getListCount((linklistsizeint*)data);
+//                bool cur_node_deleted = isMarkedDeleted(current_node_id);
+            if (collect_metrics) {
+                metric_hops++;
+                metric_distance_computations+=size;
+            }
+            dist_count+=size;
+#ifdef USE_SSE
+            _mm_prefetch((char *) (visited_array + *(data + 1)), _MM_HINT_T0);
+            _mm_prefetch((char *) (visited_array + *(data + 1) + 64), _MM_HINT_T0);
+            _mm_prefetch(data_level0_memory_ + (*(data + 1)) * size_data_per_element_ + offsetData_, _MM_HINT_T0);
+            _mm_prefetch((char *) (data + 2), _MM_HINT_T0);
+#endif
+
+            for (size_t j = 1; j <= size; j++) {
+                int candidate_id = *(data + j);
+//                    if (candidate_id == 0) continue;
+#ifdef USE_SSE
+                _mm_prefetch((char *) (visited_array + *(data + j + 1)), _MM_HINT_T0);
+                _mm_prefetch(data_level0_memory_ + (*(data + j + 1)) * size_data_per_element_ + offsetData_,
+                                _MM_HINT_T0);  ////////////
+#endif
+                if (!(visited_array[candidate_id] == visited_array_tag)) {
+                    visited_array[candidate_id] = visited_array_tag;
+
+                    char *currObj1 = (getDataByInternalId(candidate_id));
+                    dist_t dist = fstdistfunc_(data_point, currObj1, dist_func_param_);
+                    candidate_set.emplace(-dist, candidate_id,cur_hop_id+1);
+#ifdef USE_SSE
+                        _mm_prefetch(data_level0_memory_ + std::get<1>(candidate_set.top()) * size_data_per_element_ +
+                                        offsetLevel0_,  ///////////
+                                        _MM_HINT_T0);  ////////////////////////
+#endif
+                    top_candidates.emplace(dist, candidate_id);
+                    if (ground_truth_labels.count(getExternalLabel(candidate_id))) {
+                        gt_found_count++;
+                        all_gt_found = (gt_found_count == k);
+                    }
+                }
+            }
+        }
+        visited_list_pool_->releaseVisitedList(vl);
+        return top_candidates;
+    }
+
 
     void getNeighborsByHeuristic2(
         std::priority_queue<std::pair<dist_t, tableint>, std::vector<std::pair<dist_t, tableint>>, CompareByFirst> &top_candidates,
@@ -1324,6 +1428,67 @@ class HierarchicalNSW : public AlgorithmInterface<dist_t> {
     }
 
 
+    std::priority_queue<std::pair<dist_t, labeltype >>
+    searchKnnUntilAllGroundTruthFound(const void *query_data, size_t k,
+        std::set<labeltype>& ground_truth_labels, 
+        std::vector<int>& hop_id_list,
+        int& iter_count,
+        int& dist_count,
+        BaseFilterFunctor* isIdAllowed = nullptr) const {
+        std::priority_queue<std::pair<dist_t, labeltype >> result;
+        if (cur_element_count == 0) return result;
+
+        tableint currObj = enterpoint_node_;
+        dist_t curdist = fstdistfunc_(query_data, getDataByInternalId(enterpoint_node_), dist_func_param_);
+
+        for (int level = maxlevel_; level > 0; level--) {
+            bool changed = true;
+            while (changed) {
+                changed = false;
+                unsigned int *data;
+
+                data = (unsigned int *) get_linklist(currObj, level);
+                int size = getListCount(data);
+                metric_hops++;
+                metric_distance_computations+=size;
+
+                tableint *datal = (tableint *) (data + 1);
+                for (int i = 0; i < size; i++) {
+                    tableint cand = datal[i];
+                    if (cand < 0 || cand > max_elements_)
+                        throw std::runtime_error("cand error");
+                    dist_t d = fstdistfunc_(query_data, getDataByInternalId(cand), dist_func_param_);
+
+                    if (d < curdist) {
+                        curdist = d;
+                        currObj = cand;
+                        changed = true;
+                    }
+                }
+            }
+        }
+
+        std::priority_queue<std::pair<dist_t, tableint>, std::vector<std::pair<dist_t, tableint>>, CompareByFirst> top_candidates;
+        bool bare_bone_search = !num_deleted_ && !isIdAllowed;
+        if (bare_bone_search) {
+            top_candidates = searchBaseLayerSTUntilAllGroundTruthFound<true>(
+                    currObj, query_data, std::max(ef_, k), ground_truth_labels, hop_id_list,iter_count, dist_count, isIdAllowed);
+        } else {
+            top_candidates = searchBaseLayerSTUntilAllGroundTruthFound<false>(
+                    currObj, query_data, std::max(ef_, k), ground_truth_labels, hop_id_list,iter_count, dist_count, isIdAllowed);
+        }
+
+        while (top_candidates.size() > k) {
+            top_candidates.pop();
+        }
+        while (top_candidates.size() > 0) {
+            std::pair<dist_t, tableint> rez = top_candidates.top();
+            result.push(std::pair<dist_t, labeltype>(rez.first, getExternalLabel(rez.second)));
+            top_candidates.pop();
+        }
+        return result;
+    }
+
     std::vector<std::pair<dist_t, labeltype >>
     searchStopConditionClosest(
         const void *query_data,
